### DISCUSSION TITLE
How do I use enviroment variables within hostRules?

### DISCUSSION THREAD

**Discussion by jolll4** [renovatebot/renovate#25567] [Request Help] [answered]
### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, tell us what version of Renovate you run.

37.43.2

### If you're self-hosting Renovate, select which platform you are using.

Self-hosted for github, runs in circleci

### Was this something which used to work for you, and then stopped?

I never saw this working

### Wanted end result.

```
{
"token": "${{secrets.RENOVATE_TOKEN}}",
"hostRules": [
    {
      "hostType": "npm",
      "matchHost": "https://npm.pkg.github.com",
      "token": "${{secrets.RENOVATE_TOKEN}}"
    }
  ],
"npmrc": "@organization:registry=https://npm.pkg.github.com\n//npm.pkg.github.com/"
}
```
Would detect private npm packages.

### What you tried so far.

The configuration in the `renovate.json` works if I replace the token in `hostRules` with a plaintext token, but fetching the token from circle.ci enviroment variables doesn't work. The `token` outside of `hostRules` does get fetched though, so are there different rules for replacing tokens within hostRules? I've also tried using `${{secrets.GITHUB_COM_TOKEN}}`, but it doesn't work either.

### Relevant debug logs

<details><summary>Logs</summary>

```
"warnings": [
         "Failed to look up npm package @organization/package"
         ]
```

</details>

**Comment by rarkins** [comment]
Secrets are not automatically populated from env - you need to populate them on the bot config side (not renovate.json). e.g. your `config.js` could be:

```js
module.exports = {
  secrets: {
    GITHUB_COM_TOKEN: process.env.GITHUB_COM_TOKEN,
  },
};
```

**Reply by jolll4** [reply]
Should the rest of the configuration be also defined in the .js or should I be using both with only the secrets being defined in the .js? In `circleci/config.yml` the `config_file_path` was previously set to `renovate.json`, but I assume it should be pointing to `renovate-config.js` instead? Should the .json be directed to in the .js or rather in the circleci config?

**Reply by rarkins** [reply, marked as answer]
If you put this config within a JS file in your bot config then you don't even need to worry about using secrets config - just reference `process.env.FOO` directly in hostRules

If the rule applies to all repos the put it in bot hostRules config using JS.

If the rule applies to only some repos then put it in bot config secrets config and let individual repos define it in renovate.json

Do not use renovate.json for your bot config, you will just confuse yourself

**Reply by jolll4** [reply]
Thanks, got it now working with the .js. Migrated all the configs from the json and then just set 
```
npmrc: `@organization:registry=https://npm.pkg.github.com\n//npm.pkg.github.com/:_authToken=${process.env.NPM_AUTH_TOKEN}
```
